### PR TITLE
CASSANDRA-12092: fixTestAccuracy.test_simple_strategy_counters

### PR DESCRIPTION
**Pull request by stef1927**
This should hopefully fix all flakiness in this test, and possibly in other consistency tests as well, more details in [CASSANDRA-12092](https://issues.apache.org/jira/browse/CASSANDRA-12092).

### CODE REVIEW

**Comment by knifewine** [comment]
Is there any potential for these dtest changes to mask a legitimate bug?

If bug masking isn't likely, I'm +1

**Comment by knifewine** [comment]
The full suite ran fine locally for me as well.

**Comment by stef1927** [comment]
It depends what the aim of the test should be. It is true that if we disable the dynamic snitch, read repair and speculative retry then we are not testing these features, so we need to ask ourselves whether we have other tests for them.

At the moment `TestAccuracy` is testing consistency combinations that do not ensure strong consistency, and in these cases we check that the test reads the up-to-date value at least from the number of replicas corresponding to the write CL but, in order to test this, we need to change all those settings to ensure that the test always reads from the replica it contacted and, as we've seen, this is not easy. 

Maybe we should abandon this, and accept that we can only test the guarantees offered by strong consistency and that we cannot easily poke into the Cassandra internals with a distributed test. Alternatively, we could use these settings only when required, by using different tables when we don't have strong consistency, or at a minimum for other testers.

**Comment by knifewine** [comment]
@stef1927 Could it be reasonable to either a) add a second class that uses the more aggressive settings and put only the problematic tests there, or b) add a couple new helper methods to TestAccuracy for setting up slightly differently for just those tests ?

That way we could keep the general case testing basically the same and set things up better for the test(s) that really need differing settings to prevent flakiness. I think that would at least help us know that we couldn't mask problems as easily, and we'd benefit from testing a bit more configuration space.

**Comment by stef1927** [comment]
Yes it's possible to use those settings only when required, I've implemented this in 9c78f09. We could even go further and split out only the problematic combinations but that would increase the test time since for each test method we need to restart the cluster. 

I'm multiplexing the counters test against 2.1 [here](http://cassci.datastax.com/view/Dev/view/stef1927/job/stef1927-dtest-multiplex/34/) as well as running the entire consistency tests against trunk [here](http://cassci.datastax.com/view/Dev/view/stef1927/job/stef1927-dtest-multiplex/33/).

**Comment by stef1927** [comment]
The first build was aborted during the scp, restarted [here](http://cassci.datastax.com/view/Dev/view/stef1927/job/stef1927-dtest-multiplex/35/).

**Comment by stef1927** [comment]
The multiplexed build succeeded but the two builds running the full tests in `consistency_test.py` aborted twice in `TestAvailability:test_network_topology_strategy`. I've restarted a build without resource intensive tests [here](https://cassci.datastax.com/view/Dev/view/stef1927/job/stef1927-dtest-multiplex/37/). I'm not sure if it's just as simple as picking a large instance on EC2 for resource intensive tests, I normally run my multiplexed tests on Openstack.

**Comment by stef1927** [comment]
The latest build without the resource intensive tests was successful. Note that all links above are now page not found because I ran several multiplexed jobs yesterday for two other tickets.

Could I have a review of the latest changes please?

**Comment by knifewine** [comment]
If the tests are all running well with the changes, I'm +1.

File: consistency_test.py [hunk 1]
```diff
@@ -55,6 +55,29 @@ def _required_nodes(self, cl, rf_factors, dc):
...
 
+    def _get_num_nodes(self, idx, rf_factors, write_cl, read_cl):
+            """
+            Given a node index, identify to which data center we are connecting and return
+            number of nodes we write to, read from and whether R + W > N
+            """
+            nodes = [self.nodes] if isinstance(self.nodes, int) else self.nodes
+
+            dc = 0
+            for i in xrange(1, len(nodes)):
```
Comment (mambocab) [inline consistency_test.py]: Could you add a comment explaining how this loop works? It looks like it picks a datacenter, but I'm not sure how or for what.

File: consistency_test.py [hunk 2]
```diff
@@ -55,6 +55,29 @@ def _required_nodes(self, cl, rf_factors, dc):
             ConsistencyLevel.LOCAL_ONE: 1,
         }[cl]
 
+    def _get_num_nodes(self, idx, rf_factors, write_cl, read_cl):
```
Comment (mambocab) [inline consistency_test.py]: This method's name could better reflect its purpose. Maybe something about consistency? I'm not sure, it's hard for me to summarize what it does.

File: consistency_test.py [hunk 3]
```diff
@@ -55,6 +55,29 @@ def _required_nodes(self, cl, rf_factors, dc):
...
+
+            if write_cl == ConsistencyLevel.EACH_QUORUM:
+                write_nodes = sum([self._required_nodes(write_cl, rf_factors, i) for i in range(0, len(nodes))])
+            else:
+                write_nodes = self._required_nodes(write_cl, rf_factors, dc)
+
+            read_nodes = self._required_nodes(read_cl, rf_factors, dc)
+            strong_consistency = read_nodes + write_nodes > sum(rf_factors)
+
+            return write_nodes, read_nodes, strong_consistency
```
Comment (mambocab) [inline consistency_test.py]: Could you add, at the top level of this module, a `namedtuple` with fields named `write_nodes`, `read_nodes`, and `strong_consistency`, and use it for the return value of this method?

**Comment by mambocab** [comment]
Oh, feel free to ignore my comments on `_get_num_nodes`, I just noticed that you're moving existing code.

**Comment by mambocab** [comment]
Yeah, I'm also +1 if you're happy with CI results. Thanks!

**Comment by stef1927** [comment]
Thanks for the review. Indeed `_get_num_nodes()` was just moved but why not make it better, so I've implemented @mambocab suggestions in f8a3149.

I was happy with the CI results for the flaky test and for those not labeled as resource intensive, the other ones I couldn't run. Either way, after the new commit we need to re-run them so I've started a build [here](http://cassci.datastax.com/view/Dev/view/stef1927/job/stef1927-dtest-multiplex/57), I've launched it on EC2 2xlarge so hopefully the resource intensive tests will run as well.

**Comment by stef1927** [comment]
Latest CI run went well, all tests passed. Waiting for a +1 from @mambocab for the implementation of his suggestions (f8a3149) and then I think we can merge.

**Comment by mambocab** [comment]
Looks great, thank you! Merging.

**Pull request merged**